l1-contracts/contracts/bridgehub/MessageRoot.sol

```diff
@@ -120,22 +159,140 @@ contract MessageRoot is IMessageRoot, Initializable {
+            abi.encodeCall(this.sendV30UpgradeBlockNumberFromGateway, (_chainId, sentBlockNumber))
+        );
+    }
+
+    function saveV30UpgradeGatewayBlockNumberOnL1(FinalizeL1DepositParams calldata _finalizeWithdrawalParams) external {
+        bool success = proveL1DepositParamsInclusion(_finalizeWithdrawalParams, L2_MESSAGE_ROOT_ADDR);
+        if (!success) {
+            revert InvalidProof();
+        }
+
+        require(
+            BRIDGE_HUB.whitelistedSettlementLayers(_finalizeWithdrawalParams.chainId),
+            NotWhitelistedSettlementLayer(_finalizeWithdrawalParams.chainId)
+        );
+        require(block.chainid == L1_CHAIN_ID, OnlyL1());
+
+        (uint32 functionSignature, uint256 offset) = UnsafeBytes.readUint32(_finalizeWithdrawalParams.message, 0);
+        require(
+            bytes4(functionSignature) == this.sendV30UpgradeBlockNumberFromGateway.selector,
+            IncorrectFunctionSignature()
+        );
+
+        require(v30UpgradeGatewayBlockNumber == 0, V30UpgradeGatewayBlockNumberAlreadySet());
+        (uint256 chainId, ) = UnsafeBytes.readUint256(_finalizeWithdrawalParams.message, offset);
+        (uint256 receivedV30UpgradeGatewayBlockNumber, ) = UnsafeBytes.readUint256(
+            _finalizeWithdrawalParams.message,
+            offset
+        );
+        if (chainId == _finalizeWithdrawalParams.chainId) {
+            v30UpgradeGatewayBlockNumber = receivedV30UpgradeGatewayBlockNumber;
+        } else {
+            v30UpgradeChainBatchNumber[chainId] = receivedV30UpgradeGatewayBlockNumber;
+        }
+    }
+
+    function saveV30UpgradeGatewayBlockNumberOnL2(
+        uint256 _v30UpgradeGatewayBlockNumber
+    ) external onlyServiceTransactionSender {
+        v30UpgradeGatewayBlockNumber = _v30UpgradeGatewayBlockNumber;
+    }
+
+    function saveV30UpgradeChainBatchNumber(uint256 _chainId) external onlyChain(_chainId) {
+        uint256 totalBatchesExecuted = IGetters(msg.sender).getTotalBatchesExecuted();
+        require(totalBatchesExecuted > 0, TotalBatchesExecutedZero());
+        require(
+            totalBatchesExecuted != V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY &&
+                totalBatchesExecuted != V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_L1,
+            TotalBatchesExecutedLessThanV30UpgradeChainBatchNumber()
+        );
+        require(
+            v30UpgradeChainBatchNumber[_chainId] == V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_GATEWAY ||
+                v30UpgradeChainBatchNumber[_chainId] == V30_UPGRADE_CHAIN_BATCH_NUMBER_PLACEHOLDER_VALUE_FOR_L1,
+            V30UpgradeChainBatchNumberAlreadySet()
+        );
+        if (totalBatchesExecuted != 0) {
+            require(chainBatchRoots[_chainId][totalBatchesExecuted - 1] == bytes32(0), ChainBatchRootAlreadyExists(_chainId, totalBatchesExecuted));
+        }
+        v30UpgradeChainBatchNumber[_chainId] = totalBatchesExecuted;
```

Comment:
+1 is needed here (i.e. the first v30 batch is right after the latest executed one)